File: java/src/main/java/benchmarks/Revcomp.java
package benchmarks;

public class Revcomp extends Benchmark {
    private String input;
    private long resultVal;
    private static final char[] LOOKUP = new char[256];

    static {

        for (int i = 0; i < 256; i++) {
            LOOKUP[i] = (char)i;
        }

        String from = "wsatugcyrkmbdhvnATUGCYRKMBDHVN";
        String to   = "WSTAACGRYMKVHDBNTAACGRYMKVHDBN";

        for (int i = 0; i < from.length(); i++) {
            LOOKUP[from.charAt(i)] = to.charAt(i);
        }
    }

    public Revcomp() {
        resultVal = 0L;
    }

    @Override
    public String name() {
        return "Revcomp";
    }

    @Override
    public void prepare() {
        Fasta fasta = new Fasta();
        fasta.n = (int) configVal("n");
        fasta.run(0);

        String fastaResult = fasta.getResultString();

        StringBuilder seq = new StringBuilder();
        int start = 0;
        int end;

        while ((end = fastaResult.indexOf('\n', start)) != -1) {
            String line = fastaResult.substring(start, end);

            if (line.startsWith(">")) {
                seq.append("\n---\n");
            } else {
                seq.append(line);
            }

            start = end + 1;
        }

        input = seq.toString();
    }

    @Override
    public void run(int iterationId) {
        resultVal += Helper.checksum(revcompString(input));
    }

    private String revcompString(String seq) {
        int length = seq.length();
        int lines = (length + 59) / 60;
        char[] result = new char[length + lines];
        int pos = 0;

        for (int start = length; start > 0; start -= 60) {
            int chunkStart = Math.max(start - 60, 0);
            int chunkSize = start - chunkStart;

            for (int i = start - 1; i >= chunkStart; i--) {
                char c = seq.charAt(i);
                result[pos++] = LOOKUP[c];
            }

            result[pos++] = '\n';
        }

        if (length % 60 == 0 && length > 0) {
            pos--;
        }

        return new String(result, 0, pos);
    }

    @Override
    public long checksum() {
        return resultVal;
    }
}

File: scala/src/main/scala/benchmarks/Revcomp.scala
package benchmarks

class Revcomp extends Benchmark:
  private val LOOKUP = new Array[Char](256)
  private var input: String = _
  private var resultVal: Long = 0L

  private def initLookup(): Unit =
    var i = 0
    while i < 256 do
      LOOKUP(i) = i.toChar
      i += 1

    val from = "wsatugcyrkmbdhvnATUGCYRKMBDHVN"
    val to = "WSTAACGRYMKVHDBNTAACGRYMKVHDBN"

    i = 0
    while i < from.length do
      LOOKUP(from.charAt(i)) = to.charAt(i)
      i += 1

  initLookup()

  override def name(): String = "Revcomp"

  override def prepare(): Unit =
    val fasta = new Fasta()
    fasta.n = configVal("n").toInt
    fasta.run(0)
    val fastaResult = fasta.getResultString

    val seq = new StringBuilder()
    var start = 0
    var end = fastaResult.indexOf('\n', start)

    while end != -1 do
      val line = fastaResult.substring(start, end)

      if line.startsWith(">") then seq.append("\n---\n")
      else seq.append(line)

      start = end + 1
      end = fastaResult.indexOf('\n', start)

    input = seq.toString()

  private def revcompString(seq: String): String =
    val length = seq.length
    val lines = (length + 59) / 60
    val result = new Array[Char](length + lines)
    var pos = 0

    var start = length
    while start > 0 do
      val chunkStart = math.max(start - 60, 0)
      val chunkSize = start - chunkStart

      var i = start - 1
      while i >= chunkStart do
        val c = seq.charAt(i)
        result(pos) = LOOKUP(c)
        pos += 1
        i -= 1

      result(pos) = '\n'
      pos += 1
      start = chunkStart

    if length % 60 == 0 && length > 0 then pos -= 1

    String(result, 0, pos)

  override def run(iterationId: Int): Unit =
    resultVal += Helper.checksum(revcompString(input))

  override def checksum(): Long = resultVal
